File: sdk/android/tokenpay-id-sdk/src/main/java/space/tokenpay/id/internal/Telemetry.kt
package space.tokenpay.id.internal

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.serialization.json.*
import space.tokenpay.id.BuildConfig
import java.util.UUID

/**
 * Fire-and-forget funnel telemetry. Events whitelist:
 *  widget.opened, widget.step_shown, widget.step_submit_ok, widget.step_submit_error,
 *  widget.closed_success, widget.closed_cancel, widget.sign_out, widget.error_shown,
 *  widget.device_flow_started, widget.passkey_started, widget.passkey_success, widget.passkey_failed
 *
 * The SDK strips PII from properties client-side before sending. The server also applies
 * its own PII filter.
 */
internal class Telemetry(
    private val api: ApiClient,
    private val clientId: String,
    private val enabled: Boolean,
) {
    @Volatile
    private var sessionId: String = UUID.randomUUID().toString()

    private val forbiddenKeys = setOf(
        "email", "password", "code", "otp", "totp", "token", "secret", "phone",
        "card", "pan", "cvv", "ssn", "passport", "name", "full_name", "address",
        "refresh_token", "access_token", "id_token", "verifier",
    )

    fun newSession(id: String = UUID.randomUUID().toString()) {
        sessionId = id
    }

    fun track(event: String, props: Map<String, Any?> = emptyMap()) {
        if (!enabled) return
        val scrubbed = props
            .filterKeys { it.lowercase() !in forbiddenKeys }
            .mapValues { (_, v) ->
                when (v) {
                    null -> JsonNull
                    is Boolean -> JsonPrimitive(v)
                    is Number -> JsonPrimitive(v)
                    else -> JsonPrimitive(v.toString().take(200))
                }
            }
        val payload = buildJsonObject {
            put("event", event)
            put("client_id", clientId)
            put("session_id", sessionId)
            put("sdk_version", BuildConfig.TPID_SDK_VERSION)
            put("platform", "android")
            put("platform_version", android.os.Build.VERSION.RELEASE)
            put("ts", System.currentTimeMillis())
            put("properties", JsonObject(scrubbed))
        }
        GlobalScope.launch(Dispatchers.IO) {
            runCatching { api.sendTelemetry(payload) }
        }
    }
}

File: sdk/swift/Sources/TokenPayID/Internal/Telemetry.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// Fire-and-forget telemetry with client-side PII scrub.
final class Telemetry: @unchecked Sendable {
    private let api: ApiClient
    private let clientId: String
    private let enabled: Bool
    private let queue = DispatchQueue(label: "space.tokenpay.id.telemetry", qos: .utility)
    private var sessionId: String = UUID().uuidString

    private let forbiddenKeys: Set<String> = [
        "email", "password", "code", "otp", "totp", "token", "secret", "phone",
        "card", "pan", "cvv", "ssn", "passport", "name", "full_name", "address",
        "refresh_token", "access_token", "id_token", "verifier",
    ]

    init(api: ApiClient, clientId: String, enabled: Bool) {
        self.api = api; self.clientId = clientId; self.enabled = enabled
    }

    func newSession(_ id: String = UUID().uuidString) {
        queue.async { self.sessionId = id }
    }

    func track(_ event: String, properties: [String: Any] = [:]) {
        guard enabled else { return }
        queue.async {
            var scrubbed: [String: Any] = [:]
            for (k, v) in properties where !self.forbiddenKeys.contains(k.lowercased()) {
                if let s = v as? String { scrubbed[k] = String(s.prefix(200)) }
                else if v is Bool || v is Int || v is Double { scrubbed[k] = v }
                else { scrubbed[k] = String(describing: v).prefix(200).description }
            }
            let payload: [String: Any] = [
                "event": event,
                "client_id": self.clientId,
                "session_id": self.sessionId,
                "sdk_version": TpidVersion.string,
                "platform": Self.platform(),
                "platform_version": Self.platformVersion(),
                "ts": Int(Date().timeIntervalSince1970 * 1000),
                "properties": scrubbed,
            ]
            Task { await self.api.sendTelemetry(event: payload) }
        }
    }

    private static func platform() -> String {
        #if os(iOS)
        return "ios"
        #elseif os(macOS)
        return "macos"
        #elseif os(tvOS)
        return "tvos"
        #elseif os(watchOS)
        return "watchos"
        #elseif os(visionOS)
        return "visionos"
        #else
        return "apple"
        #endif
    }

    private static func platformVersion() -> String {
        #if canImport(UIKit) && !os(watchOS)
        return UIDevice.current.systemVersion
        #else
        let v = ProcessInfo.processInfo.operatingSystemVersion
        return "\(v.majorVersion).\(v.minorVersion).\(v.patchVersion)"
        #endif
    }
}
